### PR TITLE
Don't require the config.yaml for the build script

### PR DESCRIPTION
**Pull request by Pchelolo**
Some adjustments and refinements for options parsing.

cc @wikimedia/services

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -142,16 +141,34 @@ function replaceEnvVars(config) {
     });
 }
 
+function requiresConfig(options) {
+    return !options.generate && (options.start || options.test);
```
Comment (d00rman) [inline index.js:145]: AFAIK, only `start` and `test` need the config. Neither `generate` nor `build` do.

File: index.js [hunk 2]
```diff
@@ -142,16 +141,34 @@ function replaceEnvVars(config) {
...
 function main() {
     const options = getOptions();
-    let configFile = options.configFile;
-    if (!/^\//.test(configFile)) {
+
+    let configFilePath = options.configFilePath || 'config.yaml';
+    if (!/^\//.test(configFilePath)) {
         // resolve relative paths
-        configFile = path.resolve(`${process.cwd()}/${configFile}`);
+        configFilePath = path.resolve(`${process.cwd()}/${configFilePath}`);
```
Comment (d00rman) [inline index.js:154]: If `app_base_path` is set, we should honour that instead of using `CWD`

File: index.js [hunk 3]
```diff
@@ -142,16 +141,34 @@ function replaceEnvVars(config) {
...
+    }
+
+    if (requiresConfig(options) && !fs.existsSync(configFilePath)) {
+        console.error('Service config is required but not provided');
+        process.exit(1);
+    }
+
+    let config;
```
Comment (d00rman) [inline index.js]: Perhaps better to set it to an empty hash.

File: lib/docker.js [hunk 4]
```diff
@@ -542,29 +542,38 @@ function sigHandle() {
...
+        tests: options.test,
         coverage: false,
         deploy: options.buildDeploy,
         verbose: options.verbose,
         need_build: options.buildDeploy && options.force,
         review: options.sendReview,
         reshrinkwrap: options.reshrinkwrap,
         generate: options.generate,
-        path: options.basePath
+        path: options.basePath,
```
Comment (d00rman) [inline lib/docker.js]: Ugh, let's noooot pleaseeee :)

File: index.js (hunk 1 above)
Comment (Pchelolo) [reply index.js:145]: That's what is expressed here. This `generate -t` we will have both `generate` and `test` but not need the config.

File: lib/docker.js (hunk 4 above)
Comment (Pchelolo) [reply lib/docker.js]: Ugh, typo :)

File: index.js (hunk 1 above)
Comment (d00rman) [reply index.js:145]: I must be missing something. You cannot both specify a command that needs the config and one that doesn't (as only one is allowed). It follows that neither `generate` nor `build` should be taken into account here, as they don't need it.

Comment (Pchelolo) [reply index.js:145]: @d00rman See https://github.com/wikimedia/node-service-builder/pull/2/files#diff-168726dbe96b3ce427e7fedce31bb0bcR108

Comment (d00rman) [reply index.js:145]: Ah! I stand corrected!

File: index.js (hunk 2 above)
Comment (Pchelolo) [reply index.js:154]: Hm... The `app_base_path` is the path to the deployment root path. Config shouldn't really be there..

File: index.js (hunk 3 above)
Comment (Pchelolo) [reply index.js]: Not quite, we use `!config` in the docker script to verify it's provided when it's needed. Having a `TypeError` inn case some of the verification is incorrect is better then getting `undefined` for the property and silently doing something wrong.

**Pull request merged**